### PR TITLE
Feature/trustworthiness

### PR DESCRIPTION
**Pull request by humanjas**
#90 

# Introduction
Implemented the verification of a block using backtracking

# Purpose
This way you can verify a block with the database to check if it actually exists and if the public key, revoked status and iban is correct.

# Future Applicability
When you receive a new block, you will have to verify first.

# PR Reflection

### CODE REVIEW

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockControllerInterface.java [hunk 1]
```diff
@@ -122,4 +122,18 @@
      * @return if the database is empty
      */
     boolean isDatabaseEmpty();
+
+    /**
+     * backtrack method
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockControllerInterface.java]: Not descriptive enough.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 2]
```diff
@@ -158,10 +159,16 @@ public final int getLatestSeqNumber(String owner) {
     @Override
     public final List<Block> revokeBlock(Block block) {
         String owner = block.getOwner();
-        Block newBlock = BlockFactory.getBlock(REVOKE, block.getOwner(),
-                block.getOwnHash(), block.getPreviousHashChain(), block.getPreviousHashSender(),
-                block.getPublicKey(), block.getIban(), block.getTrustValue());
-        addBlock(newBlock);
+        addBlock(BlockFactory.getBlock(
+                REVOKE,
+                owner,
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Why this inconsistency?

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 3]
```diff
@@ -303,10 +318,46 @@ private Block createBlock(String owner, String contact, String publicKey, String
...
         return block;
     }
+
+    /**
+     * @inheritDoc
+     */
+    @Override
+    public Block backtrack(Block block) {
+        String previousHashSender = block.getPreviousHashSender();
+        Block loop_block = block;
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Underscores in variable names is discourages if not disallowed.

File: app/src/main/java/nl/tudelft/b_b_w/model/Block.java [hunk 4]
```diff
@@ -31,9 +31,9 @@
      * @param _trustvalue         the trust value of the contact
      * @param _isRevoked          boolean to check whether a block is revoked or not
      */
-    public Block(int _sequenceNumber, String _owner, String _ownHash, String _previousHashChain, String _previousHashSender, String _publicKey, String _iban, int _trustvalue, boolean _isRevoked) {
-        this.sequenceNumber = _sequenceNumber;
+    public Block(String _owner, int _sequenceNumber, String _ownHash, String _previousHashChain, String _previousHashSender, String _publicKey, String _iban, int _trustvalue, boolean _isRevoked) {
```
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: Why the argument shuffle?

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 2 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Because otherwise you make two get calls to the block object, while you already have the owner attribute.

File: app/src/main/java/nl/tudelft/b_b_w/model/Block.java (hunk 4 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: Because this is the original sequence, as described in the database columns.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 3 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: This has been fixed in the last commit.

**Comment by ashay123** [comment]
Can't you just put into BlockExists an extra parameter IBAN which checks if the block already exists?

**Comment by humanjas** [comment]
A block can exist in the database, but when the block is backtracked, the oldest version of the block is retrieved. This oldest version should be verified with the given block. This is the difference with the exists method.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 5]
```diff
@@ -402,6 +406,30 @@ public void testBacktrack() {
     }
 
     /**
+     * verifyTrustworthiness test
+     * Tests whether a block is trustworthy
+     */
+    @Test
+    public void testVerifyTrustworthiness() {
+        final Block newBlock = BlockFactory.getBlock(TYPE_BLOCK, owner, bc.getLatestSeqNumber(owner)+1,
```
Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: You should make a block using the methods in BlockController

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 6]
```diff
@@ -402,6 +406,30 @@ public void testBacktrack() {
...
+    }
+
+    /**
+     * verifyTrustworthiness test
+     * Tests whether a block is trustworthy
+     * Forces a false
+     */
+    @Test
+    public void testVerifyTrustworthinessFalse() {
+        final Block newBlock = BlockFactory.getBlock(TYPE_BLOCK, owner, bc.getLatestSeqNumber(owner)+1,
```
Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Again, use BlockController

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java [hunk 7]
```diff
@@ -158,6 +158,32 @@ public void testTrustValueInit() {
     }
 
     /**
+     * Test to check whether the verifyBlock() method returns the right boolean value indicating if this block is equal to the parameter block.
+     * @throws Exception Catches error when the MessageDigest
+     * gets an error.
+     */
+    @Test
+    public void verifyBlockTest() throws Exception {
+        final Block check = BlockFactory.getBlock(TYPE_BLOCK, owner, blockController.getLatestSeqNumber(owner)+1,
```
Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java]: Use BlockController

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java [hunk 8]
```diff
@@ -158,6 +158,32 @@ public void testTrustValueInit() {
...
+    /**
+     * Test to check whether the verifyBlock() method returns the right boolean value indicating if this block is equal to the parameter block.
+     * Forces a false
+     * @throws Exception Catches error when the MessageDigest
+     * gets an error.
+     */
+    @Test
+    public void verifyBlockFalseTest() throws Exception {
+        final String _owner = "NOTOWNER";
+        final Block check = BlockFactory.getBlock(TYPE_BLOCK, _owner, blockController.getLatestSeqNumber(owner),
```
Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java]: BlockController.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 5 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: The function in the BlockController is not correct, so until it has been refactored, it is discouraged to use it.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 6 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: The function in the BlockController is not correct, so until it has been refactored, it is discouraged to use it.

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java (hunk 7 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java]: The function in the BlockController is not correct, so until it has been refactored, it is discouraged to use it.

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java (hunk 8 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java]: The function in the BlockController is not correct, so until it has been refactored, it is discouraged to use it.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 5 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: So it would have to be fixed in another pull request.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 6 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: So it would have to be fixed in another pull request.

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java (hunk 7 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java]: So it would have to be fixed in another pull request.

File: app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java (hunk 8 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/model/BlockUnitTest.java]: So it would have to be fixed in another pull request.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 5 above)
Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Can you specify where the incorrectness is located?

**Comment by ashay123** [comment]
I see that for your DisplayContactList you still use ContactAdapter, which isn't made for that purpose. This way you'll get different behaviour.

**Comment by humanjas** [comment]
This issue is being addressed upon in another pull request. Please make an issue for this.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 5 above)
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: In the blockcontroller class, when you create a block, the value contact of who the block is, is needed as well. However, we are not using it anywhere else, not even to calculate the value of previousHashSender.

Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: @ymte

Comment (ymte) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: For now this is OK. There is a PR upcoming that will use the new User class for both owner and contact, which will include a refactor of the BlockFactory as well, so this will get fixed there as well.

Comment (ashay123) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Alrighty

File: app/src/main/java/nl/tudelft/b_b_w/model/Block.java [hunk 9]
```diff
@@ -129,6 +129,23 @@ public final boolean isRevoked() {
...
 
     /**
+     * Verifies whether the attributes of the given block are equal to this
+     * Attributes are isRevoked, iban and public key
+     * @param o given block
+     * @return equals or not
+     */
+    public boolean verifyBlock(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
Comment (nzarin) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: couldn't you make an equal method for this?

Comment (nzarin) [inline app/src/main/java/nl/tudelft/b_b_w/model/Block.java]: nvm
